File: java-spring/src/main/java/org/burgas/javaspring/service/contract/CrudService.java
package org.burgas.javaspring.service.contract;

import org.burgas.javaspring.dto.Request;
import org.burgas.javaspring.dto.Response;
import org.burgas.javaspring.entity.Entity;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.UUID;

@Component
public interface CrudService<R extends Request, E extends Entity, S extends Response, F extends Response> extends EntityService<E> {

    F findById(UUID id);

    List<S> findAll();

    void create(R request);

    void update(R request);

    void delete(UUID id);
}

File: kotlin-ktor/src/main/kotlin/service/CrudService.kt
package org.burgas.service

import org.burgas.database.Request
import org.burgas.database.Response
import java.util.UUID

interface CrudService<in R : Request, out S : Response, out F : Response> {

    suspend fun create(request: R): F

    suspend fun findAll(): List<S>

    suspend fun findById(id: UUID): F

    suspend fun update(request: R)

    suspend fun delete(id: UUID)
}